File: backend/internal/domain/sport.go
package domain

import (
	"math"
	"strings"
)

type SportSession struct {
	Activity   string          `json:"activity"`
	Minutes    float64         `json:"minutes"`
	DistanceKm float64         `json:"distanceKm"`
	Exercises  []SportExercise `json:"exercises"`
}
type SportExercise struct {
	Name     string  `json:"name"`
	Sets     int     `json:"sets"`
	Reps     int     `json:"reps"`
	WeightKg float64 `json:"weightKg"`
}

func (e FamilyEntry) validateSport() error {
	if e.Kind != "sport" {
		if e.Sport != nil {
			return familyInvalid("спортивные показатели доступны только в спорте")
		}
		return nil
	}
	s := e.Sport
	validNumber := func(v, max float64) bool { return !math.IsNaN(v) && !math.IsInf(v, 0) && v >= 0 && v <= max }
	if s == nil || len(e.ParticipantIDs) != 1 || !ValidFamilyDate(e.Date) || strings.TrimSpace(s.Activity) == "" || len([]rune(s.Activity)) > 80 {
		return familyInvalid("укажите участника, дату и вид спорта")
	}
	if !validNumber(s.Minutes, 1440) || s.Minutes == 0 || !validNumber(s.DistanceKm, 2000) || len(s.Exercises) > 50 {
		return familyInvalid("проверьте длительность, дистанцию и упражнения")
	}
	for _, x := range s.Exercises {
		if strings.TrimSpace(x.Name) == "" || len([]rune(x.Name)) > 100 || x.Sets < 1 || x.Sets > 100 || x.Reps < 1 || x.Reps > 10000 || !validNumber(x.WeightKg, 1000) {
			return familyInvalid("проверьте название, подходы, повторения и вес упражнения")
		}
	}
	return nil
}

File: frontend/src/domain/sport.ts
import type { FamilyEntry, SportExercise } from './family'
import { localDate } from './date'
export function sportProgress(entries: FamilyEntry[]) {
  const weeks = new Map<string, { date: string; minutes: number; count: number }>()
  type Point = SportExercise & { date: string }
  const exercises = new Map<string, { name: string; first: Point; last: Point; best: number }>()
  let minutes = 0, distance = 0
  for (const e of [...entries].sort((a, b) => a.date.localeCompare(b.date) || a.id - b.id)) {
    if (e.kind !== 'sport' || !e.sport) continue
    minutes += e.sport.minutes; distance += e.sport.distanceKm
    const day = new Date(`${e.date}T12:00:00`)
    day.setDate(day.getDate() - (day.getDay() + 6) % 7)
    const key = localDate(day), week = weeks.get(key) ?? { date: key, minutes: 0, count: 0 }
    week.minutes += e.sport.minutes; week.count++; weeks.set(key, week)
    for (const x of e.sport.exercises ?? []) {
      const key = `${e.sport.activity.trim().toLocaleLowerCase('ru')}|${x.name.trim().toLocaleLowerCase('ru')}`
      const point = { ...x, date: e.date }, old = exercises.get(key)
      exercises.set(key, { name: `${x.name} · ${e.sport.activity}`, first: old?.first ?? point, last: point, best: Math.max(old?.best ?? 0, x.weightKg) })
    }
  }
  return { minutes: Math.round(minutes * 100) / 100, distance: Math.round(distance * 100) / 100, weeks: [...weeks.values()], exercises: [...exercises.values()] }
}
